### ISSUE TITLE
Windows10 + UTF-8 ditaa diagrams bad rendering

### ISSUE THREAD

**Issue by habamax** [asciidoctor/asciidoctor-diagram#233]
Just to separate from the #150 which is cygwin related.

Ruby version: ruby 2.6.0p0 (2018-12-25 revision 66547) [x64-mingw32]

java version "1.8.0_201"
Java(TM) SE Runtime Environment (build 1.8.0_201-b09)
Java HotSpot(TM) 64-Bit Server VM (build 25.201-b09, mixed mode)

Windows language: main one is English US, second input lang is russian.

Active code page in cmd.exe is: Active code page: 65001

## test.adoc
```
== test 1
// 2019-04-22
Press the button kbd:[Alt+F4]

[ditaa, test-ditaa, png]
....
/--------\   +-------+
|cAAA    +---+Version|
|  прив  |   |   V3  |
|  Base  |   |cRED{d}|
|     {s}|   +-------+
\---+----/
....

```

![image](https://user-images.githubusercontent.com/234774/61032858-f32c5180-a3ca-11e9-8e0d-71d6f7b9b6b7.png)

**Comment by Mobe91** [comment]
Same issue here with Windows 10. I am using the asciidoctor gradle plugin (https://github.com/asciidoctor/asciidoctor-gradle-plugin) v1.5.8.1 with asciidoctor-diagram:1.5.18 and attempting to render a ditaa diagram. My input .adoc is UTF-8 encoded.

**Comment by pepijnve** [comment]
@habamax @Mobe91 Just to be sure, which build of Ruby are you using?

**Comment by Mobe91** [comment]
I am using the com.github.jruby-gradle:jruby-gradle-plugin:1.7.0 gradle plugin for ruby.

**Comment by habamax** [comment]
> 
> 
> @habamax @Mobe91 Just to be sure, which build of Ruby are you using?

ruby 2.6.0p0 (2018-12-25 revision 66547) [x64-mingw32]

**Comment by pepijnve** [comment]
> > @habamax @Mobe91 Just to be sure, which build of Ruby are you using?
> 
> ruby 2.6.0p0 (2018-12-25 revision 66547) [x64-mingw32]

Where does that binary come from? https://rubyinstaller.org?

**Comment by habamax** [comment]
> 
> 
> > > @habamax @Mobe91 Just to be sure, which build of Ruby are you using?
> > 
> > 
> > ruby 2.6.0p0 (2018-12-25 revision 66547) [x64-mingw32]
> 
> Where does that binary come from? https://rubyinstaller.org?

yes

**Comment by habamax** [comment]
Just to clarify -- all non-ditaa plantuml diagrams are ok with unicode

**Comment by Mobe91** [comment]
@habamax Just rendered a plantuml diagram without problems, so yes, it seems to affect ditaa diagrams only.

**Comment by pepijnve** [comment]
Just to keep you guys posted, I'm getting the same problem in the Ditaa UTF-8 unit test on Windows 10. Still root causing...

**Comment by pepijnve** [comment]
Bug in Ditaa right here https://github.com/stathissideris/ditaa/blob/master/src/java/org/stathissideris/ascii2image/text/TextGrid.java#L1561
They way it mangles the Strings there is just plain wrong. I'll fix that in my fork.

**Comment by pepijnve** [comment]
Would you mind giving the fix in https://github.com/asciidoctor/asciidoctor-diagram/commit/4e96e5297c75781f75b6c0c8cdba1ed9233328b8 a try?

**Comment by habamax** [comment]
> 
> 
> Would you mind giving the fix in [4e96e52](https://github.com/asciidoctor/asciidoctor-diagram/commit/4e96e5297c75781f75b6c0c8cdba1ed9233328b8) a try?

How would I do it?

What I have tried:

1. git pull this repo
2. try to use it in a "dumb" way

ruby C:/Users/maksim.kim/projects/asciidoctor-pdf/bin/asciidoctor-pdf -r C:/Users/maksim.kim/projects/asciidoctor-diagram/lib/asciidoctor-diagram.rb -a pdf-theme="../.asciidoctor-setup/habamax-theme.yml" -a pdf-fontsdir="../.asciidoctor-setup/fonts/" test.adoc

This way it didn't work for me -- unicode chars are not rendered

**Comment by pepijnve** [comment]
Ok, that should have worked. I’ll do some more testing on my end.

**Comment by habamax** [comment]
I think the issue is here https://github.com/pepijnve/ditaa/blob/de16361db241aaccd1c86c595fbf40a194455eb2/src/main/java/org/stathissideris/ascii2image/core/FileUtils.java#L60

it is either null encoding is passed or bad encoding is passed to the file reader making all unicode strings -> byte strings.

**Comment by habamax** [comment]
Or probably here https://github.com/pepijnve/ditaa/blob/de16361db241aaccd1c86c595fbf40a194455eb2/src/main/java/org/stathissideris/ascii2image/core/ProcessingOptions.java#L61

Charset.defaultCharset() will return something windows-1251 or other non-unicode encoding on windows.

https://stackoverflow.com/questions/9312816/java-platforms-default-charset-on-different-platforms

**Comment by habamax** [comment]
Indeed Charset.defaultCharset() is an issue here.

I have just used ditaamini.jar to convert a test file supplying encoding:

    java -jar ditaamini-0.12.jar -e UTF-8 test.adoc

and it worked!

![test adoc](https://user-images.githubusercontent.com/234774/61659290-d7c91c80-accf-11e9-9434-533103c1f466.png)

**Comment by pepijnve** [comment]
Now that's odd, because that's essentially exactly what the ditaa integration is doing. There are a couple of parts to this:

1. At https://github.com/asciidoctor/asciidoctor-diagram/blob/master/lib/asciidoctor-diagram/ditaa/extension.rb#L57 the ditaa code is sent to Java#send_request
That's going to send an HTTP request to the Java processing daemon.
2. https://github.com/asciidoctor/asciidoctor-diagram/blob/master/lib/asciidoctor-diagram/util/java.rb#L15 the source string is encoded in its native encoding, whatever that may be. The encoding is sent over the wire as well via the `Content-Type` header (`text/plain; charset=...`).
3. The receiving end https://github.com/asciidoctor/asciidoctor-diagram-java/blob/master/ditaa/src/main/java/org/asciidoctor/diagram/ditaa/Ditaa.java#L53 takes the bytes from the HTTP request body and the encoding from the Content-Type header and passes that on to Ditaa.

The bug I fixed earlier (https://github.com/pepijnve/ditaa/commit/fed5158168bd79c6a8514f394ff797e4ff372e49) was that Ditaa would take the decoded String, dump it to bytes using the default encoding and then decode again using the encoding given on the command line. That simply makes no sense and so I removed that bit of logic.

I'll try to reproduce the issue with asciidoctor-pdf exactly the way you had it set up. Maybe there's something else I'm missing.

**Comment by habamax** [comment]
>     2\. https://github.com/asciidoctor/asciidoctor-diagram/blob/master/lib/asciidoctor-diagram/util/java.rb#L15 the source string is encoded in its native encoding, whatever that may be. The encoding is sent over the wire as well via the `Content-Type` header (`text/plain; charset=...`).

https://github.com/asciidoctor/asciidoctor-diagram/blob/4e96e5297c75781f75b6c0c8cdba1ed9233328b8/lib/asciidoctor-diagram/util/java.rb#L16

I can see ascii encoding to be set there...

**Comment by pepijnve** [comment]
That's intentional :) The header part of an HTTP request is US_ASCII encoded. Switches to BINARY at https://github.com/asciidoctor/asciidoctor-diagram/blob/4e96e5297c75781f75b6c0c8cdba1ed9233328b8/lib/asciidoctor-diagram/util/java.rb#L44 before dumping the code contents.

I've been trying to reproduce the problem here but no luck so far. Before my change I was getting the same problem; after the change things work correctly regardless of the default encodings I set on both Ruby and the JVM.

**Comment by habamax** [comment]
> I've been trying to reproduce the problem here but no luck so far. Before my change I was getting the same problem; after the change things work correctly regardless of the default encodings I set on both Ruby and the JVM.

Could you try to use UTF-8 instead of `Charset.defaultCharset()` in ditaa default options just to check if it would work?

**Comment by pepijnve** [comment]
The default encoding is never used since the exact encoding (`UTF-8`) is passed to Ditaa as a command line argument. In other words, that's already being done.

Let me add some debug logging to the Java code so we can figure out what's actually happening.

**Comment by habamax** [comment]
> 
> 
> The default encoding is never used since the exact encoding (`UTF-8`) is passed to Ditaa as a command line argument. In other words, that's already being done.
> 
> Let me add some debug logging to the Java code so we can figure out what's actually happening.

Ok, got it.

And https://github.com/asciidoctor/asciidoctor-diagram/blob/4e96e5297c75781f75b6c0c8cdba1ed9233328b8/lib/asciidoctor-diagram/ditaa/extension.rb#L10

encoding is not in this OPTIONS, could it be that further it was not correctly sent to ditaa?

**Comment by pepijnve** [comment]
It's not passed as a CLI argument. It gets extracted from the HTTP request Content-Type header. See https://github.com/asciidoctor/asciidoctor-diagram-java/blob/master/ditaa/src/main/java/org/asciidoctor/diagram/ditaa/Ditaa.java#L47

**Comment by habamax** [comment]
> 
> 
> It's not passed as a CLI argument. It gets extracted from the HTTP request Content-Type header. See https://github.com/asciidoctor/asciidoctor-diagram-java/blob/master/ditaa/src/main/java/org/asciidoctor/diagram/ditaa/Ditaa.java#L47

Ok, you know it better :)

Just in case
https://github.com/asciidoctor/asciidoctor-diagram-java/blob/f028dfddec2293698a5b69085044251c71b6c229/ditaa/src/main/java/org/asciidoctor/diagram/ditaa/Ditaa.java#L47

Is charset of content type is correctly set up then?

**Comment by habamax** [comment]
Looks like it should be UTF-8 looking at the constants

**Comment by habamax** [comment]
And the last one :)

Could it be that ByteArrayInputStream corrupts the UTF strings?

https://github.com/asciidoctor/asciidoctor-diagram-java/blob/f028dfddec2293698a5b69085044251c71b6c229/ditaa/src/main/java/org/asciidoctor/diagram/ditaa/Ditaa.java#L70

**Comment by pepijnve** [comment]
ByteArrayInputStream is simply going to return the bytes verbatim. It's when they get converted to a String that the actual decoding happens. Ditaa does that at https://github.com/pepijnve/ditaa/blob/master/src/main/java/org/stathissideris/ascii2image/text/TextGrid.java#L1267 which uses the configured encoding.

**Comment by habamax** [comment]
Ok, so strategically placed printlns probably should help then.

**Comment by pepijnve** [comment]
Is that something you would be comfortable doing? I don't have time to work on this right now, maybe this evening. The way I've debugged this stuff in the past is by starting up the Java daemon with a debugger manually and then patching the Ruby code to connect to the already running server instead of spawning a new one. Basically just hardcoding the port of the running daemon at https://github.com/asciidoctor/asciidoctor-diagram/blob/master/lib/asciidoctor-diagram/util/java_socket.rb#L23 instead of reading it from the launched processes stdout.

**Comment by habamax** [comment]
1. I don't have anything java dev related on my machines
2. I don't think I understand what should be exactly done when you describe "daemon... connect to running server" etc :)
3. I don't use ditaa at all (probably because it didn't work for me with unicode in the first place quite some time ago)

That said, I can for sure test the changes provided but setup correct dev environment and do even the simplest debugging is smth I would avoid doing at least for the upcoming 2-3 months

**Comment by habamax** [comment]
@pepijnve looks like you somehow fixed this issue :)

![image](https://user-images.githubusercontent.com/234774/74809897-48d0c100-52ff-11ea-9ab6-4490a7e99285.png)

![image](https://user-images.githubusercontent.com/234774/74809941-64d46280-52ff-11ea-9ff6-d958bc48bc2a.png)

**Comment by habamax** [comment]
it is most probably your new implementation of server communication which landed in 2.0.0 (and was fixed for windows in 2.0.1)

@Mobe91 could you also check if it works for you?
